src/structure.js

```diff
@@ -169,13 +169,19 @@ Structure.prototype.reference = function (path) {
   if (isCursor(path) && path._keyPath) {
     path = path._keyPath;
   }
-  var self = this, pathId = pathString(path);
-  var listenerNs = self._pathListeners[pathId];
-  var cursor = this.cursor(path);
 
-  var changeListener = function (newRoot, oldRoot, changedPath) { cursor = self.cursor(path); };
-  var referenceListeners = [changeListener];
-  this._pathListeners[pathId] = !listenerNs ? referenceListeners : listenerNs.concat(changeListener);
+  path = valToKeyPath(path) || [];
+
+  var self = this,
+      listenerNs = getListenerNs(self._pathListeners, path),
+      cursor = this.cursor(path),
+      changeListener = function() {
+        cursor = self.cursor(path);
+      };
+
+  if (!listenerNs[0]) {
+    listenerNs.push(changeListener);
```

Comment:
It seems to me, that if you has on `getListenerNs(self._pathListeners, path)` you can get a object literal in return, not an array. This means `.push` will be undefined? As `self._pathListeners` (object) is passed into `utils.deepGet`, which returns the passed into value (which is `self._pathListeners`). And the value from `utils.deepGet` will get returned from `getListenerNs`. So `listenerNs` will be an object and `({}).push => undefined`, no?